src/ucs/memory/rcache_vfs.c

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright (C) 2021 NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <ucs/config/global_opts.h>
+#include <ucs/datastruct/queue.h>
+#include <ucs/sys/compiler.h>
+#include <ucs/sys/math.h>
+#include <ucs/sys/string.h>
+#include <ucs/vfs/base/vfs_cb.h>
+#include <ucs/vfs/base/vfs_obj.h>
+#include "rcache_int.h"
+#include "rcache_vfs.h"
+
+
+#define UCS_RCACHE_FORMAT_STRING ("%12s:%9zi\n")
+
+
+static void ucs_rcache_vfs_read_inv_q_length(void *obj,
+                                             ucs_string_buffer_t *strb,
+                                             void *arg_ptr, uint64_t arg_u64)
+{
+    ucs_rcache_t *rcache = obj;
+    size_t rcache_inv_q_length;
+
+    ucs_spin_lock(&rcache->lock);
+    rcache_inv_q_length = ucs_queue_length(&rcache->inv_q);
+    ucs_spin_unlock(&rcache->lock);
+
+    ucs_string_buffer_appendf(strb, "%zu\n", rcache_inv_q_length);
+}
+
+static void ucs_rcache_vfs_read_gc_list_length(void *obj,
+                                               ucs_string_buffer_t *strb,
+                                               void *arg_ptr, uint64_t arg_u64)
+{
+    ucs_rcache_t *rcache = obj;
+    unsigned long rcache_gc_list_length;
+
+    ucs_spin_lock(&rcache->lock);
+    rcache_gc_list_length = ucs_list_length(&rcache->gc_list);
+    ucs_spin_unlock(&rcache->lock);
+
+    ucs_string_buffer_appendf(strb, "%lu\n", rcache_gc_list_length);
+}
+
+static void ucs_rcache_vfs_calc_regions_distribution(ucs_rcache_t *rcache,
+                                                     size_t *rcache_dist,
+                                                     size_t num_bins)
+{
+    ucs_config_thresholds_t *thresholds = &ucs_global_opts.rcache_thresholds;
+    ucs_rcache_region_t *region;
+    size_t region_size, i;
+
+    for (i = 0; i < num_bins; ++i) {
+        rcache_dist[i] = 0;
+    }
+
+    ucs_spin_lock(&rcache->lru.lock);
+
+    ucs_list_for_each(region, &rcache->lru.list, lru_list) {
```

Comment:
if we show distributions by power-of-2, we could keep 64-entry array of counters and update it when add/remove region, instead of scanning all regions every time (which is too time consuming even for gathering statistics IMO)

Reply:
Each created rcache has its own distribution in this implementation. Do you propose to have total distribution among all registration caches? If not so, we should have an array for each cache. Otherwise, we need a synchronization to access to the array.
It is better to have array of counters for each rcache, because number of caches usually is small, IMO.

Reply:
as discussed offline - can keep a variable size array per rcache, something like log2(min/max)